src/cargo/ops/cargo_install.rs

```diff
@@ -0,0 +1,329 @@
+            });
+            let examples = candidates.iter().filter(|cand| {
+                cand.targets().iter().filter(|t| t.is_example()).count() > 0
+            });
+            let pkg = match try!(one(binaries, |v| multi_err("binaries", v))) {
+                Some(p) => p,
+                None => {
+                    match try!(one(examples, |v| multi_err("examples", v))) {
+                        Some(p) => p,
+                        None => return Err(human("no packages found with \
+                                                  binaries or examples")),
+                    }
+                }
+            };
+            return Ok((pkg.clone(), Box::new(source)));
+
+            fn multi_err(kind: &str, mut pkgs: Vec<&Package>) -> String {
+                pkgs.sort_by(|a, b| a.name().cmp(b.name()));
+                format!("multiple packages with {} found: {}", kind,
+                        pkgs.iter().map(|p| p.name()).collect::<Vec<_>>()
+                            .join(", "))
+            }
+        }
+    }
+}
+
+fn one<I, F>(mut i: I, f: F) -> CargoResult<Option<I::Item>>
+    where I: Iterator,
+          F: FnOnce(Vec<I::Item>) -> String
+{
+    match (i.next(), i.next()) {
+        (Some(i1), Some(i2)) => {
+            let mut v = vec![i1, i2];
+            v.extend(i);
+            Err(human(f(v)))
+        }
+        (Some(i), None) => Ok(Some(i)),
+        (None, _) => Ok(None)
+    }
+}
+
+fn check_overwrites(dst: &Path,
+                    pkg: &Package,
+                    filter: &ops::CompileFilter,
+                    prev: &CrateListingV1) -> CargoResult<()> {
+    let check = |name| {
+        let name = format!("{}{}", name, env::consts::EXE_SUFFIX);
+        if fs::metadata(dst.join(&name)).is_err() {
+            return Ok(())
+        }
+        let mut msg = format!("binary `{}` already exists in destination", name);
+        if let Some((p, _)) = prev.v1.iter().find(|&(_, v)| v.contains(&name)) {
+            msg.push_str(&format!(" as part of `{}`", p));
+        }
+        Err(human(msg))
+    };
+    match *filter {
+        CompileFilter::Everything => {
+            // If explicit --bin or --example flags were passed then those'll
+            // get checked during cargo_compile, we only care about the "build
+            // everything" case here
+            if pkg.targets().iter().filter(|t| t.is_bin()).next().is_none() {
+                return Err(human("specified package has no binaries"))
+            }
+
+            for target in pkg.targets().iter().filter(|t| t.is_bin()) {
+                try!(check(target.name()));
+            }
+        }
+        CompileFilter::Only { bins, examples, .. } => {
+            for bin in bins.iter().chain(examples) {
+                try!(check(bin));
+            }
+        }
+    }
+    Ok(())
+}
+
+fn read_crate_list(path: &Path) -> CargoResult<CrateListingV1> {
+    let metadata = path.join(".crates.toml");
```

Comment:
My concern is that this is
1.) A general enough tool to tempt people to apply it to system-wide use (i.e., it supports arbitrary destinations via `--root` and such)
2.) The metadata is sufficiently useful that there is an incentive for others to parse it
3.) The metadata as specified is *unnecessarily* hostile to system PMs (i.e., doing it this way doesn't seem to provide much in the way of benefit over ways that would be less hostile)
4.) Simply *not* installing it at all for the system PM is a potential hazard due to (2) (as well as Cargo itself if `--root` is passed)

Those make it such that once the problematic way is used, it's going to be very difficult to use a way that is *not*.

As for a (simple!) solution that's far less problematic, simply having each crate's metadata go in `${ROOT}/some_dir/${CRATE_HASH}.crate.toml` would do the trick - it'd be a very limited amount of added complexity (iterate the directory, merge the parsed data) in exchange for something that won't need a redesign later.

(Personally, I'd suggest `some_dir` be `lib/cargo`, by analogy with pkgconfig.)